### ISSUE TITLE
HMR could be faster

### ISSUE THREAD

**Issue by orta** [artsy/reaction#211]
I pulled out some stats on our webpack 

https://ortastuff.s3.amazonaws.com/data/webpack_stats.json - which you can download, and open with http://webpack.github.io/analyse/

---

I don't have too much internal understanding of webpack as my focus has mainly been on [metro bundler](https://github.com/facebook/metro-bundler/). 

We see most of the wait inside ` 92% chunk asset optimization`. So what can we learn from this?

We have 2 chunks, I expect this chunk (id: 1) is the [storybooks code](http://webpack.github.io/analyse/#chunk/1) 

![screen shot 2017-08-01 at 15 26 56](https://user-images.githubusercontent.com/49038/28843068-dec91e0c-76cd-11e7-9c45-b2b8ece9d56a.png)

We don't want this to be re-created on an incremental build.

This second one is our app, 

![screen shot 2017-08-01 at 15 31 42](https://user-images.githubusercontent.com/49038/28843270-87c4a1ca-76ce-11e7-841c-1740a06bb0eb.png)

You can find the modules in here: http://webpack.github.io/analyse/#module/81

![screen shot 2017-08-01 at 15 31 42](https://user-images.githubusercontent.com/49038/28843288-9979fc58-76ce-11e7-9466-230f32745d30.png)

--

When you look at the data that gets sent on a hot update, it's tiny:

![screen shot 2017-08-01 at 15 33 06](https://user-images.githubusercontent.com/49038/28843324-ba6f3e0a-76ce-11e7-9e5a-514066925dbf.png)

It's just taking a bunch of time to get there

**Comment by orta** [comment]
I pushed a branch with what it took to generate this stuff:

```sh
git clone https://github.com/artsy/reaction-force.git
git checkout stats
yarn install
yarn storybooks
```

Then use `curl http://localhost:9001/stats.json > stats.json` to download the stats JSON.

**Comment by orta** [comment]
Options for further digging:

* https://github.com/amireh/happypack

**Comment by orta** [comment]
This HMR is what I care about: https://twitter.com/niieani/status/894630819192557568

**Comment by orta** [comment]
https://medium.com/webpack/typescript-webpack-super-pursuit-mode-83cc568dea79

**Comment by damassi** [comment]
@orta - this might provide some visual insight in terms of efficient organization: https://github.com/webpack/webpack/issues/690

**Comment by orta** [comment]
Something worth highlighting around build times is that our editor + commit linter + CI provide all the build time checking, if we're talking 2s -> 0.Xs transformations it could be worth dropping that aspect

**Comment by l2succes** [comment]
tried switching to webpack v3.0.0 with ts-loader + happypack to no avail.
Here's the `webpack.config.js` we used:

```javascript
const env = require("dotenv")
const fs = require("fs")
const path = require("path")
const sharify = require("./sharify")

const webpack = require("webpack")
const webpackMerge = require("webpack-merge")
const { CheckerPlugin } = require("awesome-typescript-loader")
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin")
const HappyPack = require("happypack")

/**
 * Write out a file that stubs the data that’s normally shared with the client through the `sharify` module. This file
 * is then replaced in the product of webpack where normally the actual `sharify` module would be loaded.
 */
const { METAPHYSICS_ENDPOINT } = env.config().parsed
const sharifyPath = sharify({ METAPHYSICS_ENDPOINT })

/**
 * Only use HMR plugin in dev mode
 */
let plugins = [new CheckerPlugin()]
if (process.env.NODE_ENV === "development") {
  plugins.push(new webpack.HotModuleReplacementPlugin())
  plugins.push(
    new HappyPack({
      id: "ts",
      threads: 2,
      loaders: [
        {
          path: "ts-loader",
          query: { happyPackMode: true },
        },
      ],
    })
  )
  plugins.push(new ForkTsCheckerWebpackPlugin({ checkSyntacticErrors: false }))
}

// A mix of  the base from Emission's webpack setup, and the simple config for
// storybooks: https://storybook.js.org/configurations/custom-webpack-config/

module.exports = {
  devtool: "#inline-source-map", // Otherwise getting errors about e.g. `Relay` not being defined.
  resolve: {
    extensions: [".js", ".jsx", ".ts", ".tsx"],
    alias: {
      sharify: sharifyPath.replace(/\.js$/, ""),
    },
  },
  module: {
    rules: [
      { test: /\.json$/, loader: "json-loader" },
      { test: /\.tsx?$/, loader: "babel-loader!happypack/loader?id=ts" },
    ],
  },
  plugins: plugins,
}
```

**Comment by damassi** [comment]
@l2succes - What was the exact outcome?

**Comment by orta** [comment]
The time to do see a HMR change was more or less the same 6 to 7 seconds

**Comment by l2succes** [comment]
@damassi  We were getting times in the 5-6secs ballpark for one small change. Here's a sample of times we were getting after changing one file.

![screen shot 2017-09-06 at 4 11 58 pm](https://user-images.githubusercontent.com/296775/30132418-4bdb4602-931e-11e7-8540-32cd2a680a67.png)

**Comment by damassi** [comment]
Have you looked into moving vendor assets into a DLL? Its a little bit of setup, but at my last gig where we were facing a similar situation it vastly improved compile times and was well worth the effort to figure out how to implement:

https://webpack.js.org/plugins/dll-plugin/
https://gist.github.com/MoOx/64ff089e3d0ed0a48705cf330eb20111d

Another option is when Typescript compilations moves into Babel, you can take advantage of `babel-loader`'s `cacheDirectory`, which is off by default: https://github.com/babel/babel-loader#options.

**Comment by orta** [comment]
Nope 👍

**Comment by l2succes** [comment]
No we haven't, definitely worth giving it a try though. I'll do those and see if we see any improvements.

**Comment by damassi** [comment]
Awesome -- that example gist should help. I remember it being a bit of a headache to finally get working but once it was done I was like "hell yeah....." -- 7-10 second build times down to around 1 - 2.  It was pretty dramatic.

**Comment by orta** [comment]
Also, if you get this working well @l2succes - write a blog post

**Comment by l2succes** [comment]
Perfect! Sounds like what we need @damassi.

**Comment by l2succes** [comment]
@orta sure if works out. It might take me a lil while, but I'll ask you or Ash for some assist :)

**Comment by damassi** [comment]
Also, you can swap out `inline-source-map` with `eval` or `cheap-eval-source-map` in the [storybooks config](https://github.com/damassi/reaction/blob/master/.storybook/webpack.config.js#L29) during development, and if you're running into tricky issues swap it back to something more precise. Storybook dev isn't tricky runtime application stuff so we'll probably be fine 99% of the time. In the past I've used `yarn start:sourcemap` to enable slower compilation when I needed it, defaulting to fast mode. 

Initial build time:

```sh
[at-loader] Ok, 2.016 sec.
webpack built 6785dd903fd04027d330 in 9341ms
Storybook started on => http://localhost:9001/

webpack building...
 94% asset optimization
[at-loader] Checking started in a separate process...

[at-loader] Ok, 0 sec.
webpack built e2633caba652ec91183e in 938ms
```

And then subsequent rebuilds hover around: 

```sh
[at-loader] Ok, 1.217 sec.
webpack built f45ea04907335f0af951 in 2054ms
```

**Comment by damassi** [comment]
Sorry one more thought! That streaming text (chunks, etc) compile log output could be optimized I bet. I would recommend looking into https://www.npmjs.com/package/progress-bar-webpack-plugin and https://www.npmjs.com/package/friendly-errors-webpack-plugin to clean things up a bit.

**Comment by orta** [comment]
Pretty sure the node version is to make sure it's the same as force's

**Comment by alloy** [comment]
Yeah Force’s version. We could decide to just run that version on CI if Node 8 yields significant speed, but I do worry about people having a hard time figuring out that CI might be failing because of Node version.

**Comment by alloy** [comment]
There’s context here, but it won’t open in a helpful way on the iOS app, so not sure what it is https://github.com/artsy/reaction/pull/119

**Comment by damassi** [comment]
Let's bump them up. We'll get a nice perf boost on Forces side, too :)

**Comment by damassi** [comment]
I could be totally off here but I think we're past the point where we'll get odd failures due to versioning and if anything it's worth a try; our tech choices are pretty standard, and to test is just a single commit that we could easily flip back.
